chapter_01/final_exercise_01.md

```diff
@@ -20,27 +19,142 @@
+
 5. **Q:** What purpose does the `core-site.xml` file serve?
+5. **A:** The purpose of the core site is to configure mostly IO related things and also the filesystem and network addresses.
+some examples include: 
+ - [x] hadoop.security.authentication - either simple (no auth) or kerberos, sets whether to use authentication and which type
+ - [x] hadoop.tmp.dir - the base directory where hadoop can store temporary files
+ - [x] hadoop.security.dns.nameserver - the hostname / ip addr of the dns server which the kerberos service nodes use to their own hostname for kerberos login
+ - [x] io.bytes.per.checksum - the number of bytes to use per checksum, must be less then io.file.buffer.size
+ - [x] io.file.buffer.size - sets the buffer size for seqeunce files during read and write opperations (sequence files are binary files containing key-value paris which were created to solve some of the small file problems in hadoop by having multiple files inside one block file)
 
 ### Chapter 2: Hadoop Distributed File System (HDFS)
 
 6. **Q:** What is HDFS?
+6. **A:** HDFS is the Hadoop distributed system which is basically like a regular file system but it is distributed between multiple computers.
+HDFS is built for fault tolorence meaning, it was built with the idea that hardware failures are a norm and happen a lot so hdfs uses replication and redundency to keep the system highly available and fault tolorent.
+HDFS is the primary storage of the Hadoop applications, based on java and which contains data nodes and name nodes, the *name nodes* are the nodes which monitors all the data stored, replica counts and the amount of data nodes which are alive and the FS structure.
+the *data node* is the node that stores the data itself in the form of blocks which have a max size of a preconfigured amount, default is usually 128 MB but sometimes can be 64 or 256 MB by default.
+It is a rack aware system which means that it knows the physical topology of the network and the connected nodes which it uses to enhence fault tolorence by splitting replicas to different racks (an example with 3 replicas and 2 racks, hadoop will store one replica on the same rack with the name node and 2 more replicas on different racks in a different location / switch).
+
+provides High availability using 2 Name nodes with one being active and one being a standby, in case the active node will fail.
+
+Hdfs also has journal nodes which store changes to the file system (file addition, deletion and more) which is an edit log which is used to build the FSImage and the file system structure in the ram.
+
+Hdfs knows that "moving data is more expensive than moving computation" so it has minimal data movement and makes most of the computations in the same node that stores it's data.
```

Comment:
HDFS doesn't handle computations. The computation is a layer above, HDFS only handles IO operations and is agnostic to what is being done on it. If someone needs to read a file it will refer it to the closest data node, but no more.